Chimera - `carthage` support.

laxmorek opened the pull request:
### Short description 📝

Working on https://github.com/tuist/tuist/issues/1674

Adds possibility to install `carthage` dependencies defined in `Tuist/Dependencies.swift` using `tuist dependencies fetch` and `tuist dependencies update` commands.

### Comments 🗒

- In `.gitignore` of `app_with_framework_and_tests_and_dependencies` I added `Tuist/Dependencies/*`. I don't want to track generated files.
- <del>Frameworks are rebuilding with every run of `tuist dependencies *`. Im going to fix it in the next PR.</del> I have done it 
- I didn't add documentation and changelog changes. Feature isn't production ready. 
- Let me know if documentation in the code is understandable. Please suggest me improvements if any need. 

### Implementation 👩‍💻👨‍💻

- [x] Implement `tuist dependencies fetch/update` for `carthage`
- [x] Add unit tests
- [x] Add new fixture `app_with_framework_and_tests_and_dependencies`
- [x] Add acceptance tests

pepibumur commented on Sources/TuistDependencies/DependenciesManagerInteractors/Interactors/CarthageFrameworksInteractor.swift:
```diff
@@ -0,0 +1,121 @@
+
+        let versionFiles = try readVersionFiles(at: buildDirectoryPath)
+
+        var alreadyCopiediOSFramworks = Set<String>()
```
I'm not sure I fully understand why something like this is necessary. Is it possible to have a Carthage Build directory with the same framework more than once? When would that scenario happen?

pepibumur commented on Sources/TuistDependencies/DependenciesManagerInteractors/CarthageInteractor.swift:
```diff
@@ -0,0 +1,139 @@
+        let platoforms: Set<Platform> = dependencies
+            .reduce(Set<Platform>()) { platforms, dependency in platforms.union(dependency.platforms) }
+
+        try fileHandler.inTemporaryDirectory { temporaryDirectoryPath in
```
Unless I've missed it, I don't see logic to remove any existing frameworks that are no longer needed. Let's say we have the following scenario:
- Frameworks A, B, and C exist locally
- You build frameworks 1, 2, and C.

I'd expect 1 and 2 to be removed, A and B to show up in the directory, and C to only be built if it's necessary. I've seen that doing the latter is part of a future improvement, but what about the selection of 1 and 2?

Here's my suggestion on how to implement it:
1. Get the list of existing frameworks.
2. Get the list of built frameworks after running the Carthage command.
3. Diff them and determine which ones should be deleted and which ones copied.

laxmorek replied:
It is more tricky. What if frameworks from `Diff them and determine which ones should be deleted` have been installed via `Cocoapods` or `SPM`? I should delete them just because they don't exist in `Get the list of built frameworks after running the Carthage command.`

Right now Im going to provide simply diffing logic (like you proposed) assuming that we have only `carthage` but it will need to be improved in the future.

laxmorek replied:
- I refactored `CarthageFrameworksInteractor`
- I introduced `Tuist/Directory/graph.json`
- I added `CarthageFrameworksInteractorTests.test_save_remove_existing`

For sure diffing will need improvements. I think it will be moved outside `CarthageInteractor` for better diffing dependencies between different dependencies managers. 

In this PR Im introducing "initial" solution.

pepibumur commented on Sources/TuistDependencies/DependenciesManagerInteractors/Interactors/CarthageFrameworksInteractor.swift:
```diff
@@ -0,0 +1,121 @@
+        var alreadyCopiedtvOSFramworks = Set<String>()
+        var alreadyCopiedwatchOSFramworks = Set<String>()
+
+        try versionFiles.forEach {
```
I'd revisit the way this closure is written, if you look at each code block there's a lot of repetition that I think could be avoided writing the code differently.

pepibumur reviewed (changes requested):
Hey @laxmorek. Despite all my comments, I think the work is going in the right direction and the work is very aligned with how we do things in the project. Congratulations fur such a great contribution 👏

facumenzella commented on Sources/TuistDependencies/DependenciesManagerInteractors/Builders/CartfileContentBuilder.swift:
```diff
@@ -0,0 +1,67 @@
+            return #"github "\#(name)" ~> \#(version)"#
+        case let .upToNextMinor(version):
+            return #"github "\#(name)" ~> \#(version)"#
+        case let .range(fromVersion, toVersion):
```
Nice one 💡

laxmorek replied:
- I added `CartfileContentGenerator.validate()`method
- I added `CartfileContentGenetatorTests.test_build_invalid_dependencies()` test method

laxmorek commented on Sources/TuistDependencies/DependenciesManagerInteractors/Builders/CarthageCommandBuilder.swift:
```diff
@@ -0,0 +1,93 @@
+
+// MARK: - Carthage Command Builder
+
+final class CarthageCommandBuilder {
```
- I refactored `CarthageCommandBuilder ` into `CarthageCommandGenerator`
- I refactored `CartfileContentBuilder` into `CartfifleContentGenerator`
- I updated unit tests

laxmorek commented on features/dependencies.feature:
```diff
@@ -0,0 +1,18 @@
+Feature: Install dependencies Tuist.
+
+    Scenario: The project is an application with framework and tests and Dependencies.swift (app_with_framework_and_tests_and_dependencies)
```
1st scenario runs `tuist dependencies fetch`
2nd scenario runs `tuist dependencies update`

Should I run both commands in one scenario?

laxmorek commented on Sources/TuistDependencies/DependenciesManagerInteractors/CarthageInteractor.swift:
```diff
@@ -0,0 +1,139 @@
+        self.carthageFrameworksInteractor = carthageFrameworksInteractor
+    }
+
+    #warning("TODO: The hardes part here will be knowing whether we need to recompile the frameworks")
```
I removed the comment. Im not creating an issue. Im already working on it. I don't forget about this then. :)

laxmorek commented on Sources/TuistDependencies/DependenciesManagerInteractors/Builders/CartfileContentBuilder.swift:
```diff
@@ -0,0 +1,67 @@
+    }
+}
+
+private extension CarthageDependency {
```
- I moved it.
- I also added `CarthageDependencyTests` in `TuistCoreTests`.

laxmorek commented:
@pepibumur Thank you for your review. I adjusted PR to your suggestion. Could you revisit it?

luispadron commented on Sources/TuistCore/Models/CarthageDependency.swift:
```diff
@@ -1,13 +1,73 @@
+    var description: String {
+        switch self {
+        case let .rangeRequirementNotSupported(dependencyName, fromVersion, toVersion):
+            return "\(dependencyName) in version between \(fromVersion) and \(toVersion) can not be installed. Carthage do not support versions range requirement in Cartfile."
```
```suggestion
            return "\(dependencyName) in version between \(fromVersion) and \(toVersion) can not be installed. Carthage does not support versions range requirement in Cartfile."
```

luispadron commented on Sources/ProjectDescription/Dependency.swift:
```diff
@@ -3,10 +3,10 @@ import Foundation
 
     /// Type of requirement for the given dependency
-    let requirement: Dependency.Requirement
+    public let requirement: Dependency.Requirement
```
Wonder if we could make this requirement specific to the Dependency manager. That way you wouldn't need to throw an error for carthage range versions. 

It feels odd that the API allows this.

laxmorek replied:
yea I have the same feeling

What do you think @pepibumur @facumenzella  ?

pepibumur replied:
Sounds good! I'm always in favor of using the type-system to validate the manifest. Linting should be the very last resource.

ferologics commented:
Reminder about the Carthage `--use-netrc` flag. It's essential when installing dependencies such as [MapBox](https://docs.mapbox.com/ios/maps/overview/):
> Install the SDK with carthage update --platform iOS --use-netrc
> Always use the --use-netrc flag when running carthage bootstrap

pepibumur commented on Sources/TuistCore/Models/CarthageDependency.swift:
```diff
@@ -1,13 +1,73 @@
+        case let .upToNextMinor(version):
+            return #"github "\#(name)" ~> \#(version)"#
+        case let .range(fromVersion, toVersion):
+            throw CarthageDependencyError.rangeRequirementNotSupported(dependencyName: name, fromVersion: fromVersion, toVersion: toVersion)
```
As @luispadron mentioned, I think we can have a custom Requirement type for Carthage and leverage the type-system to prevent users from using ranges in their manifest files.

pepibumur commented on Sources/TuistDependencies/Carthage/CarthageInteractor.swift:
```diff
@@ -0,0 +1,133 @@
+    var description: String {
+        switch self {
+        case .carthageNotFound:
+            return "Carthage was not found either in Bundler nor in the environment."
```
Bundler is a tool to distribute Ruby Gems :). I'd rather say something like:

```
Carthage was not found in the environment. It's possible that the tool is not installed or hasn't been exposed to your environment.
``

laxmorek replied:
Of course. Sorry. Copy-paste issue. Im fixing it.

pepibumur commented on Sources/TuistDependencies/Carthage/Generators/CartfileContentGenerator.swift:
```diff
@@ -0,0 +1,66 @@
+
+// MARK: - Cartfile Content Generating Error
+
+enum CartfileContentGeneratorError: FatalError, Equatable {
```
Once you introduce a requirement type specific to Carthage, I think you can remove this file entirely because we no longer need to validate.

pepibumur commented on Sources/TuistDependencies/DependenciesController.swift:
```diff
@@ -1,59 +1,48 @@
 
-    public func update(at _: AbsolutePath) throws {
-        logger.notice("Start updating depednencies.")
+        try carthageInteractor.install(dependenciesDirectory: dependenciesDirectory, method: method, dependencies: dependencies.carthageDependencies)
```
I think we can log a section right before this line saying that we are starting to update the Carthage dependencies.

pepibumur commented on Sources/TuistDependencies/Models/Graph.swift:
```diff
@@ -0,0 +1,33 @@
+import Foundation
+import TuistCore
+
+struct Graph: Codable, Equatable {
```
I wouldn't call this one `Graph` to prevent conflicts with the main graph. What about calling it `DependenciesGraph`?

laxmorek replied:
Im removing `Graph` from this PR.

pepibumur commented on Sources/TuistDependencies/Models/Graph.swift:
```diff
@@ -0,0 +1,33 @@
+    let iOSDependencies: [String]
+    let tvOSDependencies: [String]
+    let macOSDependencies: [String]
+    let watchOSDependencies: [String]
```
I think the structure of this graph has been specifically designed for Carthage. This requires further thinking to make sure that the structure works for the other package managers that we plan to support: CocoaPods and SPM. Would it be ok to leave this out of the PR and move the discussion to an issue?

pepibumur commented on Sources/TuistDependencies/Carthage/Interactors/CarthageFrameworksInteractor.swift:
```diff
@@ -0,0 +1,103 @@
+        self.fileHandler = fileHandler
+    }
+
+    public func copyFrameworks(carthageBuildDirectory: AbsolutePath, dependenciesDirectory: AbsolutePath) throws {
```
As mentioned in another, I wouldn't know any graph-related work in this PR. The graph generation should be something generic enough to work with other dependency managers.

pepibumur commented on Sources/TuistDependencies/Carthage/Interactors/CarthageFrameworksInteractor.swift:
```diff
@@ -0,0 +1,103 @@
+        try fileHandler.copy(from: fromPath, to: toPath)
+    }
+
+    private func deleteDirectory(at path: AbsolutePath) throws {
```
This method is only used from once place and it's 3 lines of code so I think we can remove it and move the code to the place where the method is called from.

pepibumur commented on Sources/TuistDependencies/Carthage/Interactors/CarthageFrameworksInteractor.swift:
```diff
@@ -0,0 +1,103 @@
+
+    // MARK: - Helpers
+
+    private func buildCarthagePlatfromBuildsDirectory(carthageBuildDirectory: AbsolutePath, platform: Platform) -> AbsolutePath {
```
I'd suggest to add the following to `Platform.swift`

```swift
enum Platform {
  var carthageDirectory: String {
    // Return the directory
  }
}
```

facumenzella commented on Sources/TuistSupport/Constants.swift:
```diff
@@ -27,6 +27,13 @@ public struct Constants {
         public static let swiftDoc = "swift-doc"
     }
 
+    public struct DependenciesDirectory {
+        public static let name = "Dependencies"
+        public static let graphName = "graph.json"
+        public static let lockfilesDirectoryName = "Lockfiles"
+        public static let cartfileResolvedName = "Cartfile.resolved"
+    }
```
Can we convert it to an enum? Since everything is static, we should avoid people going `DependenciesDirectory()`

facumenzella replied:
I see that everything is a struct though 😅

laxmorek replied:
I have followed convention in this file. 

`Constants.Vendor`, `Constants.DerivedDirectory` etc

Im agree with you but I don't want to introduce a mess here.

laxmorek replied:
Basically we should convert every `struct` in this file to `enum`. Should I do it in this PR? 🤔

laxmorek commented:
### Update 🙂

#### Dependencies.swift

From now dependency's requirements depends on dependency manager. There is also no longer common `Dependency` model. I created `CarthageDependency` and in the future we will provide `CocoaPodDependency` and `SwiftPackageManagerDependency`. API of `Dependencies.swift` has been changed.

Before:

```swift
import ProjectDescription

let dependencies = Dependencies([
    .carthage(name: "Alamofire/Alamofire", requirement: .exact("5.0.4"), platforms: [.macOS]),
    .carthage(name: "Swinject/Swinject", requirement: .exact("2.7.1"), platforms: [.macOS]),
])
```

Now:

```swift
import ProjectDescription

let dependencies = Dependencies(
    carthage: [
        .init(name: "Alamofire/Alamofire", requirement: .exact("5.0.4"), platforms: [.macOS]),
        .init(name: "Swinject/Swinject", requirement: .exact("2.7.1"), platforms: [.macOS]),
    ]
)
```

#### Tuist/Dependencies directory

New structure of `Tuist/Dependencies` directory:

```bash
Tuist
  |- Dependencies
    |- graph.json # not part of this PR 
    |- Lockfiles
      |- Carthage.resolved
      |- Podfile.lock # not part of this PR 
      |- Package.resolved # not part of this PR 
    |- Carthage
      |- Build
      |- Alamofire
        |- iOS
          |- Alamofire.framework
        |- tvOS
          |- Alamofire.framework
    |- Cocoapods # not part of this PR 
      |- RxSwift
    |- SwftPackageManager # not part of this PR 
      |- Moya
```

#### Frameworks handling

- I added a logic that removes frameworks from Tuist/Dependencies` if they are no longer needed.
- I added a logic that prevents rebuilding of frameworks. 

#### Carthage command 

- I added `--use-netrc` flag for `carthage bootstrap` and `carthage update`.

#### Others

- I updated changelog
- **I didn't update documentation**

luispadron reviewed (commented):
I haven't taken an in-depth look at the code since I lack the context here to make that useful but looking at the public API this feature is looking really nice! 

Thanks for making the API more explicit and less possible to hit errors!

pepibumur commented on Sources/TuistLoader/Models+ManifestMappers/Dependencies+ManifestMapper.swift:
```diff
@@ -0,0 +1,40 @@
+import TuistCore
+import TuistSupport
+
+extension TuistCore.Dependencies {
```
Nitpick: I'd have it in 2 files to make them easier to find.
```
Dependencies+ManifestMapper.swift
CarthageDependency+ManifestMapper.swift
```

pepibumur commented on Sources/TuistKit/Services/DependenciesService.swift:
```diff
@@ -0,0 +1,39 @@
+    }
+
+    func run(path: String?, method: InstallDependenciesMethod) throws {
+        logger.info("Start installing dependencies.", metadata: .section)
```
Nitpick: `Fetching dependencies`
_(to be consistent with the command, which is named `fetch`)_

pepibumur commented on Sources/TuistKit/Services/DependenciesService.swift:
```diff
@@ -0,0 +1,39 @@
+        let dependencies = try dependenciesModelLoader.loadDependencies(at: path)
+        try dependenciesController.install(at: path, method: method, dependencies: dependencies)
+
+        logger.info("Successfully installed dependencies.", metadata: .success)
```
Nitpick: `Dependencies were fetched successfully`

pepibumur commented on Sources/TuistDependencies/Carthage/CarthageInteractor.swift:
```diff
@@ -0,0 +1,204 @@
+        case .carthageNotFound:
+            return "Carthage was not found in the environment. It's possible that the tool is not installed or hasn't been exposed to your environment."
+        case .cartfileNotFound:
+            return "Cartfile was not found after Cartage installation."
```
Typo: Cartage

pepibumur commented on Sources/TuistDependencies/Carthage/CarthageInteractor.swift:
```diff
@@ -0,0 +1,204 @@
+        case .cartfileNotFound:
+            return "Cartfile was not found after Cartage installation."
+        case .buildDirectoryNotFound:
+            return "Carthage/Build directory was not found after Cartage installation."
```
Typo: Cartage

pepibumur commented on Sources/TuistDependencies/Carthage/CarthageInteractor.swift:
```diff
@@ -0,0 +1,204 @@
+    }
+
+    public func install(dependenciesDirectory: AbsolutePath, method: InstallDependenciesMethod, dependencies: [CarthageDependency]) throws {
+        logger.info("Start installing Carthage dependencies.", metadata: .subsection)
```
Nitpick: `Installing Carthage dependencies`.

pepibumur commented on Sources/TuistDependencies/Carthage/CarthageInteractor.swift:
```diff
@@ -0,0 +1,204 @@
+    // MARK: - Installation
+
+    private func prepareForInstallation(pathsProvider: PathsProvider, dependencies: [CarthageDependency]) throws {
+        // copy build directory from previous run if exist
```
Why do we need to do this?

pepibumur commented on Sources/TuistDependencies/Carthage/CarthageInteractor.swift:
```diff
@@ -0,0 +1,204 @@
+        // save `Cartfile.resolved`
+        try copyFile(from: pathsProvider.temporaryCarfileResolvedPath, to: pathsProvider.destinationCarfileResolvedPath)
+        // save build directory
+        try copyDirectory(from: pathsProvider.temporaryCarthageBuildDirectory, to: pathsProvider.destinationCarthageBuildDirectory)
```
Why do we need to copy the `Build` directory?

pepibumur commented on Sources/TuistDependencies/Carthage/CarthageInteractor.swift:
```diff
@@ -0,0 +1,204 @@
+        }
+
+        // copy `Cartfile.resolved` from previous run if exist
+        if fileHandler.exists(pathsProvider.destinationCarfileResolvedPath) {
```
If I'm not wrong, I don't think Carthage uses the `Cartfile.resolved` file to determine what frameworks need to be fetched. If that's the case _(which I'd recommend validating)_, I don't think we need to copy this file.

laxmorek replied:
You have right. Im removing this step.

pepibumur commented on Sources/TuistDependencies/Carthage/Interactors/CarthageFrameworksInteractor.swift:
```diff
@@ -0,0 +1,81 @@
+    }
+
+    public func copyFrameworks(carthageBuildDirectory: AbsolutePath, destinationDirectory: AbsolutePath) throws {
+        let alreadyInstalledFrameworks: Set<String> = try getAlreadyInstalledFrameworkNames(at: destinationDirectory)
```
Nitpick: I'd remove `get` from the name of the method and call it just: `existingFrameworkNames`.

laxmorek replied:
CarthageFrameworksInteractor has been deleted.

pepibumur commented on Sources/TuistDependencies/Carthage/Interactors/CarthageFrameworksInteractor.swift:
```diff
@@ -0,0 +1,81 @@
+            .reduce(into: Set<String>()) { $0.insert($1) }
+    }
+
+    private func copyDirectory(from fromPath: AbsolutePath, to toPath: AbsolutePath) throws {
```
This method is in multiple files so I'd recommend extracting it into `FileHandler` and naming it:
```swift
func copyDirectory(from fromPath: AbsolutePath, to toPath: AbsolutePath, force: Bool) throws
```

laxmorek replied:
`CarthageFrameworksInteractor` has been deleted.

pepibumur commented on Sources/TuistDependencies/Carthage/Interactors/CarthageFrameworksInteractor.swift:
```diff
@@ -0,0 +1,81 @@
+    public func copyFrameworks(carthageBuildDirectory: AbsolutePath, destinationDirectory: AbsolutePath) throws {
+        let alreadyInstalledFrameworks: Set<String> = try getAlreadyInstalledFrameworkNames(at: destinationDirectory)
+
+        try Platform.allCases.forEach { platform in
```
I'd recommend getting rid of this class altogether and update the `CarthageInteractor` to delete all the Carthage frameworks from the destination directory and move the new ones that have been built.  We can improve the logic in the future using the Cartfile.resolved to do some diffing.

pepibumur commented on features/dependencies.feature:
```diff
@@ -0,0 +1,22 @@
+        Given that tuist is available
+        And I have a working directory
+        Then I copy the fixture app_with_framework_and_tests_and_dependencies into the working directory
+        Then tuist fetches dependencies
```
Note that if the file is these acceptance tests are not added to the GitHub [workflow](https://github.com/tuist/tuist/blob/main/.github/workflows/tuist.yml#L58) they won't run on CI.

laxmorek replied:
<img width="932" alt="Zrzut ekranu 2020-12-16 o 17 01 38" src="https://user-images.githubusercontent.com/4774319/102374834-172d9280-3fc2-11eb-8bb2-45245cc63139.png">

It is already added. Am I missing something?

pepibumur commented on fixtures/app_with_framework_and_tests_and_dependencies/Tuist/Dependencies.swift:
```diff
@@ -0,0 +1,8 @@
+let dependencies = Dependencies(
+    carthage: [
+        .init(name: "Alamofire/Alamofire", requirement: .exact("5.0.4"), platforms: [.macOS]),
+        .init(name: "Swinject/Swinject", requirement: .exact("2.7.1"), platforms: [.macOS]),
```
With the goal of minimizing the time it takes for the acceptance test to run, I'd reduce this list to only one framework. Unless the second one adds a new scenario, which I don't think it does.

pepibumur commented on fixtures/ios_app_with_helpers/Tuist/Dependencies.swift:
```diff
@@ -1,5 +1,8 @@
-])
+let dependencies = Dependencies(
+    carthage: [
+        .init(name: "Alamofire/Alamofire", requirement: .exact("5.0.4"), platforms: [.macOS]),
```
Looking at the syntax, I think we can make the `init` private, and expose:

```swift
static func github(name: String, requirement: Requirement, platforms: [Platform]) -> CarthageDependency {}
```

Then this code would be more expressive:

```swift
let dependencies = Dependencies(
  carthage: [
    .github(name: "Alamofire/Alamofire", requirement: .exact("5.0.4"), platforms: [.macOS])
  ]
)
```

laxmorek commented:
### Update 🙂

#### Dependencies.swift

I changed `ProjectDescription.Dependencies` model to an `enum`.  Now we can deal with more convince API:

```swift
import ProjectDescription

let dependencies = Dependencies([
    .carthage(name: "Alamofire/Alamofire", requirement: .exact("5.0.4"), platforms: [.macOS]),
    .carthage(name: "Swinject/Swinject", requirement: .exact("2.7.1"), platforms: [.macOS]),
])
```

#### Tuist/Dependencies directory

New structure of `Tuist/Dependencies` directory:

```bash
Tuist
  |- Dependencies
    |- graph.json # not part of this PR 
    |- Lockfiles
      |- Carthage.resolved
      |- Podfile.lock # not part of this PR 
      |- Package.resolved # not part of this PR 
    |- Carthage # stores content of `Carthage/Build` generated by `carthage`
      |- iOS
         |- Alamofire.framework
      |- tvOS
         |- Alamofire.framework
    |- Cocoapods # not part of this PR 
      |- RxSwift
    |- SwftPackageManager # not part of this PR 
      |- Moya
```

#### Frameworks handling

- I added a logic that copies `Carthage/Build` into `Tuist/Dependencies/Carthage`
- I added more unit tests
- I fixed acceptance tests

pepibumur commented:
> I changed ProjectDescription.Dependencies model to an enum. Now we can deal with more convince API:


Note that Carthage supports indicating multiple sources: `github`, `gitlab`, `local`. However, the interface proposed here doesn't account for that and assumes that the dependencies are fetched from GitHub. Here's a proposal:

```swift
import ProjectDescription

let dependencies = Dependencies([
    .carthage(path: "Alamofire/Alamofire", source: .github, requirement: .exact("5.0.4"), platforms: [.macOS]),
    .carthage(path: "Swinject/Swinject", source: .gitlab, requirement: .exact("2.7.1"), platforms: [.macOS]),
    .carthage(path: "Swinject/Swinject", source: .local, platforms: [.macOS]),
])
```

> New structure of Tuist/Dependencies directory:

This looks great :👏 😌

pepibumur commented on Sources/ProjectDescription/Dependencies/Dependencies.swift:
```diff
@@ -0,0 +1,10 @@
+import Foundation
+
+public struct Dependencies: Codable, Equatable {
```
This is a user-facing model and therefore should be documented. Otherwise, developers won't get documentation when editing the manifest.

pepibumur commented on Sources/ProjectDescription/Dependencies/Dependency+CarthageRequirement.swift:
```diff
@@ -0,0 +1,69 @@
+import Foundation
+
+public extension Dependency {
+    enum CarthageRequirement: Codable, Equatable {
```
Nitpick: Would you mind moving this to `Dependency.swift` and remove the extension? Not having to jump between files to glance through the models related to `Dependency` results in a better developer experience.

laxmorek replied:
You suggested me opposite thing [here](https://github.com/tuist/tuist/pull/2060#discussion_r539189717). Could you explain me a difference?

pepibumur replied:
In that case, we only have the extension option because the models are defined in `TuistCore`, and the loading method that we are adding lives in `TuistLoader`. In this case, everything is part of the same framework, so using an extension seems a bit unnecessary and the compiler will be happier about it because it'll be able to parse things faster.

laxmorek replied:
ok I have here a few options: 

1.

```swift
public enum Dependency: Codable, Equatable {
    case carthage(name: String, requirement: CarthageRequirement, platforms: Set<Platform>)

    // MARK: - CarthageRequirement

    enum CarthageRequirement: Codable, Equatable {
        case exact(Version)
        case upToNextMajor(Version)
        case upToNextMinor(Version)
        case branch(String)
        case revision(String)

        private enum Kind: String, Codable {
            case exact
            case upToNextMajor
            case upToNextMinor
            case branch
            case revision
        }

        private enum CodingKeys: String, CodingKey {
            case kind
            case version
            case branch
            case revision
        }

        public init(from decoder: Decoder) throws {
            //
        }

        public func encode(to encoder: Encoder) throws {
            //
        }
    }
}
```

2.

```swift
public enum Dependency: Codable, Equatable {
    case carthage(name: String, requirement: CarthageRequirement, platforms: Set<Platform>)
}

// MARK: - CarthageRequirement

extension Dependency {
    enum CarthageRequirement: Codable, Equatable {
        case exact(Version)
        case upToNextMajor(Version)
        case upToNextMinor(Version)
        case branch(String)
        case revision(String)
        
        private enum Kind: String, Codable {
            case exact
            case upToNextMajor
            case upToNextMinor
            case branch
            case revision
        }

        private enum CodingKeys: String, CodingKey {
            case kind
            case version
            case branch
            case revision
        }

        public init(from decoder: Decoder) throws {
            //
        }
        
        public func encode(to encoder: Encoder) throws {
            //
        }
    }
}
```

3. Similar to `ProjectDescription.Setup`.

```swift
public enum Dependency: Codable, Equatable {
    enum CarthageRequirement: Codable, Equatable {
        case exact(Version)
        case upToNextMajor(Version)
        case upToNextMinor(Version)
        case branch(String)
        case revision(String)
    }
    
    case carthage(name: String, requirement: CarthageRequirement, platforms: Set<Platform>)
}

extension Dependency.CarthageRequirement {
    private enum Kind: String, Codable {
        case exact
        case upToNextMajor
        case upToNextMinor
        case branch
        case revision
    }
    
    private enum CodingKeys: String, CodingKey {
        case kind
        case version
        case branch
        case revision
    }
    public init(from decoder: Decoder) throws {
        //
    }
    
    public func encode(to encoder: Encoder) throws {
        //
    }
}
```

4. I have started to think about moving `CarthageRequirement` outside of `Dependency`. 

4a - `CarthageRequirement` and `Dependency` same file
4b- `CarthageRequirement` and `Dependency` separate files

```swift
public enum Dependency: Codable, Equatable {
    case carthage(name: String, requirement: CarthageRequirement, platforms: Set<Platform>)
}

public enum CarthageRequirement: Codable, Equatable {
    case exact(Version)
    case upToNextMajor(Version)
    case upToNextMinor(Version)
    case branch(String)
    case revision(String)
    
    private enum Kind: String, Codable {
        case exact
        case upToNextMajor
        case upToNextMinor
        case branch
        case revision
    }
    
    private enum CodingKeys: String, CodingKey {
        case kind
        case version
        case branch
        case revision
    }
    public init(from decoder: Decoder) throws {
        //
    }
    
    public func encode(to encoder: Encoder) throws {
        //
    }
}
```

pepibumur commented on Sources/ProjectDescription/Dependencies/Dependency+CarthageRequirement.swift:
```diff
@@ -0,0 +1,69 @@
+import Foundation
+
+public extension Dependency {
+    enum CarthageRequirement: Codable, Equatable {
```
This model is also user-facing and therefore should be documented.

pepibumur commented on Sources/ProjectDescription/Dependencies/Dependency.swift:
```diff
@@ -0,0 +1,39 @@
+import Foundation
+
+public enum Dependency: Codable, Equatable {
```
All public elements here should be documented.

pepibumur commented on Tests/ProjectDescriptionTests/Dependencies/Dependency+CarthageRequirementTests.swift:
```diff
@@ -0,0 +1,31 @@
+
+@testable import ProjectDescription
+
+final class DependencyCarthageRequirementTests: XCTestCase {
```
After removing the extension remember to move these tests into `DependencyTests`.

pepibumur commented on Sources/TuistKit/Services/DependenciesService.swift:
```diff
@@ -0,0 +1,45 @@
+
+    func run(path: String?, method: InstallDependenciesMethod) throws {
+        switch method {
+        case .fetch: logger.info("Start fetching dependencies.", metadata: .section)
```
When we are printing logs, we are talking to the users and therefore we should refer to the second person and not the infinitive. I suggest changing the logs here to something along the lines of:

```
We are starting to fetch/update the dependencies.
```

pepibumur commented on Sources/TuistKit/Services/DependenciesService.swift:
```diff
@@ -0,0 +1,45 @@
+        let path = self.path(path)
+
+        let dependencies = try dependenciesModelLoader.loadDependencies(at: path)
+        try dependenciesController.install(at: path, method: method, dependencies: dependencies)
```
Nitpick: If we are calling the actions "fetch" and "update", I think this method should be called "fetch" to be consistent. These are subtle things, but when there are many of them, they might compromise the readability of the code. Also, I'd recommend not doing anything if the user decides to update because update is not supported yet. If they try to update and we do fetch instead we are doing something that they don't expect us to do:

```swift
switch method {
  case .fetch:
    logger.info("We are starting to fetch the dependencies", metadata: .section)
    let dependencies = try dependenciesModelLoader.loadDependencies(at: path)
    try dependenciesController.fetch(at: path, method: method, dependencies: dependencies)
  case .update:
    logger.info("Updating dependencies is not yet supported.")
}
```

laxmorek replied:
I think that `update` is supported. `CarthageCommandGenerator` can build `carthage update` command.

pepibumur replied:
Right, the generator can generate the command, but can I update dependencies by running `tuist dependencies update`?

laxmorek replied:
I guess so. 

`tuist dependencies update` triggers `carthage update`. `carthage update` fetches the newest dependencies that matches requirements. 

`tuist dependencies fetch` triggers `carthage bootstrap`. carthage bootstrap` fetches dependencies that match `Cartfile.resolved`. If  `Cartfile.resolved` doesn't exist,  the newest dependencies will be fetched.

laxmorek replied:
I have just realized that I removed coping `Cartfile.resolved` from previous run (https://github.com/tuist/tuist/pull/2060#discussion_r539231951). I have to restore it to make sure that `tuist dependencies fetch` will fetch locked versions.

laxmorek replied:
@pepibumur after a few tests I have decided to mark it as `supported yet` like you suggested. 

I will revisit it the upcoming PRs.

pepibumur commented on features/dependencies.feature:
```diff
@@ -0,0 +1,20 @@
+Feature: Install dependencies Tuist.
```
Since we don't want the `Tuist/Dependencies` directory to be added to the Git repository by default, I'd recommend extending the git ignore for the `app_with_framework_and_tests_and_dependencies` fixture, to ignore the `Tusit/Dependencies` directory. Moreover, I'd update the gitignore of the template that we use to scaffold new projects to include `Tuist/Dependencies`. You can add a new section [below this one](https://github.com/tuist/tuist/blob/main/Templates/Gitignore.stencil#L65):

```
### Tuist managed dependencies ###
Tuist/Dependencies
```

laxmorek replied:
1. I have already added `Tuist/Dependencies` to `.gitignore` for `app_with_framework_and_tests_and_dependencies`. You can check it [here](https://github.com/tuist/tuist/blob/b8fb0bc95d46bf34a59c9cc1c605967afdf67a50/fixtures/app_with_framework_and_tests_and_dependencies/.gitignore#L66). Of course I can change `### Generated ###` title to `### Tuist managed dependencies ###`.
2. I am not sure about update the `gitignore` of the template. It should be user decision if he/she want to push dependencies to git.

pepibumur replied:
> I am not sure about update the gitignore of the template. It should be user decision if he/she want to push dependencies to git.
 
I'm inclined git ignoring them by default. If they disagree with Tuist's stance, they can delete the line from the `.gitignore` file.

pepibumur commented on Sources/TuistDependencies/Carthage/CarthageInteractor.swift:
```diff
@@ -0,0 +1,195 @@
+
+    public func install(dependenciesDirectory: AbsolutePath, method: InstallDependenciesMethod, dependencies: [CarthageDependency]) throws {
+        switch method {
+        case .fetch: logger.info("Start fetching Carthage dependencies.", metadata: .section)
```
Similar comment in regards to the tense of the logs. You are telling the user something; if we use infinitive, the user doesn't feel like we are telling them something.

pepibumur commented on Sources/TuistDependencies/Carthage/CarthageInteractor.swift:
```diff
@@ -0,0 +1,195 @@
+    private func prepareForInstallation(pathsProvider: PathsProvider, dependencies: [CarthageDependency]) throws {
+        // copy build directory from previous run if exist
+        if fileHandler.exists(pathsProvider.destinationCarthageDirectory) {
+            try copyDirectory(from: pathsProvider.destinationCarthageDirectory, to: pathsProvider.temporaryCarthageBuildDirectory)
```
Why are we doing this? How does Carthage use any existing frameworks?

laxmorek replied:
I need it to prevent frameworks recompiling.

pepibumur commented on Sources/TuistDependencies/Carthage/CarthageInteractor.swift:
```diff
@@ -0,0 +1,195 @@
+
+// MARK: - PathsProvider
+
+private extension CarthageInteractor {
```
I don't think we need extensions here. We can rename `PathsProvider` to `CarthagePathsProvider`, make it `private`, and move it to the root of the file.

laxmorek replied:
I need it to prevent frameworks recompiling.

ferologics commented:
Hey, just my two cents on this as passive observer:

> `.carthage(path: "Alamofire/Alamofire", source: .github, requirement: .exact("5.0.4"), platforms: [.macOS])`

Feels to me like the `path` and `source` could also be passed in as one parameter:

`.carthage(source: .github(path: "Alamofire/Alamofire", requirement: .exact("5.0.4"), platforms: [.macOS])`

Just an idea 🙂

pepibumur commented:
I like thaT API @ferologics. What do you think @laxmorek ?

laxmorek commented:
yea, I like it too :)

laxmorek commented:
I found (here)[https://github.com/Carthage/Carthage/blob/master/Documentation/Artifacts.md#origin] supported origins. 

`github`, `git`, `binary`

@pepibumur I can't find info about `gitlab`, `local`.

fortmarek commented:
> I found [here](https://github.com/Carthage/Carthage/blob/master/Documentation/Artifacts.md#origin) supported origins: `github`, `git`, `binary`
> 
> I can't find info about `gitlab`, `local`.

Yep, `gitlab` is not supported (at least not as github). To install carthage dependency from gitlab, you can do `git "git@gitlab.url:Repo"`. Same for `local` -> if you want to import local carthage dependency, you can prepend it with `file://`. I'd add only cases that are supported in the Cartfile, though and users can fallback to `git` when necessary

laxmorek commented:
### Update 

- I split `DependenciesService` into `DependenciesFetchService` and `DependenciesUpdateService`
- I marked `tuist dependencies update` as `not supported`
- I added `CarthageOrigin` to manifest
- I added missing `atLeast` to  `CarthageRequirement`
- I added more documentation
- I fixed unit tests and acceptance tests
- I rebased on `main`
- I fixed changelog.

pepibumur commented on Tests/ProjectDescriptionTests/DependencyTests.swift:
```diff
@@ -0,0 +1,55 @@
+
+@testable import ProjectDescription
+
+final class DependencyTests: XCTestCase {
```
This should inherit from `TuistUnitTestCase`

pepibumur commented on Tests/ProjectDescriptionTests/DependenciesTests.swift:
```diff
@@ -0,0 +1,14 @@
+
+@testable import ProjectDescription
+
+final class DependenciesTests: XCTestCase {
```
This should inherit from `TuistUnitTestCase`

pepibumur commented on Tests/TuistCoreTests/Models/CarthageDependencyTests.swift:
```diff
@@ -0,0 +1,119 @@
+@testable import TuistCore
+@testable import TuistSupportTesting
+
+final class CarthageDependencyTests: TuistTestCase {
```
Since this is a unit test, it should inherit from `TuistUnitTestCase`.

pepibumur reviewed (approved):
Brilliant work @laxmorek. I think we are ready to merge here after addressing the final comments.
